rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java

```diff
@@ -2537,38 +2535,67 @@ private int determineParenthesisLevel(int childLineNumber, int parentLineNumber,
         int start = sourceLineNumberOffsets[startingLineNumber - 1] + startingColumn - 1;
         int end = sourceLineNumberOffsets[endingLineNumber - 1] + endingColumn - 1;
 
-        // Determine level of parentheses by going through the source
+        // Determine level of parentheses by going through the source, don't count parentheses in comments, string literals and regexes
         int count = 0;
+        Delimiter delimiter = null;
         for (int i = start; i < end; i++) {
-            if (source.charAt(i) == '(') {
-                count++;
-            } else if (source.charAt(i) == ')') {
-                count--;
+            if (delimiter == null) {
+                delimiter = getDelimiter(i);
+                if (delimiter == null) {
+                    if (source.charAt(i) == '(') {
+                        count++;
+                    } else if (source.charAt(i) == ')') {
+                        count--;
+                    }
+                }
+            } else if (delimiter.end.equals(source.substring(i, Math.min(i + delimiter.end.length(), source.length())))) {
+                i += delimiter.end.length() - 1; // skip the next chars for the rest of the delimiter
+                delimiter = null;
             }
         }
         return Math.max(count, 0);
     }
 
-    private @Nullable String getDelimiter() {
+    private @Nullable Delimiter getDelimiter(int cursor) {
         String maybeDelimiter = source.substring(cursor, Math.min(cursor + 3, source.length()));
         if (maybeDelimiter.startsWith("$/")) {
-            return "$/";
+            return DOLLAR_SLASHY_STRING;
         } else if (maybeDelimiter.startsWith("\"\"\"")) {
-            return "\"\"\"";
+            return TRIPLE_DOUBLE_QUOTE;
         } else if (maybeDelimiter.startsWith("'''")) {
-            return "'''";
+            return TRIPLE_SINGLE_QUOTE;
         } else if (maybeDelimiter.startsWith("~/")) {
-            return "~/";
+            return PATTERN_OPERATOR;
+        } else if (maybeDelimiter.startsWith("//")) {
+            return SINGLE_LINE_COMMENT;
+        } else if (maybeDelimiter.startsWith("/*")) {
+            return MULTILINE_COMMENT;
         } else if (maybeDelimiter.startsWith("/")) {
-            return "/";
+            return SLASHY;
         } else if (maybeDelimiter.startsWith("\"")) {
-            return "\"";
+            return DOUBLE_QUOTE;
         } else if (maybeDelimiter.startsWith("'")) {
-            return "'";
+            return SINGLE_QUOTE;
         }
         return null;
     }
 
+    @RequiredArgsConstructor
+    enum Delimiter {
```

Comment:
We could lift this field even to the `G` class. The GString class has also a delimiter field:

```java
final class GString implements G, Statement, Expression {
  UUID id;
  Space prefix;
  Markers markers;
  String delimiter;
  //...
```

Even the printer does already some magic to make it work:

```java
public J visitGString(G.GString gString, PrintOutputCapture<P> p) {
     //...   
     if ("$/".equals(delimiter)) {
       p.append("/$");   // <- which could then now be modeled like `delimiter.end`
     } else {
       p.append(delimiter);
     }
```

Problem would be we break current synchronised LSTs, so I did not implement this. Any ideas here @knutwannheden?

Reply:
I don't think it is worth to change the model. We could instead provide a utility method to print a string using certain delimiters.

Reply:
What about moving the `Delimiter` enum to a separate file? Then we could just transform above code to:

```java
public J visitGString(G.GString gString, PrintOutputCapture<P> p) {
  //...   
  Delimiter delimiter = Delimiter.of(gString.getDelimiter());
  p.append(delimiter.open);
  visit(gString.getStrings(), p);
  p.append(delimiter.close);
  //..
}
```

I think that's a nice middle ground of little improvement without altering too much.

Reply:
Yes, that sounds like an option. Let's add it to an `internal` package. I don't really see recipes needing this.